gateway/README.md

```diff
@@ -0,0 +1,377 @@
+
+The Success JSON structure provides a uniform structure for expressing normal responses using a structure similar to the Error JSON structure used to render errors. The structure provides an enumerated set of codes and associated HTTP statuses (see Errors below) along with a message.
+
+The Success JSON structure has the following format. The results tag is optional and appears when the response contains one or more resources.
+```
+{
+  "success": {
+    "status": <http-status-code>,
+    "code": <enumerated-error-code>,
+    "message": <message-text>
+  },
+  "results": <service-response>
+}
+```
+
+`results` content follows the [Google model](https://cloud.google.com/apis/design/standard_methods): an object is returned for Get, Create and Update operations and list of objects for List operation.
+
+To allow compatibility with existing systems, the results tag name can be changed to a service-defined tag. In this way the success data becomes just a tag added to an existing structure.
+
+#### Examples:
+
+Response with no results:
+```
+{
+  "success": {
+    "status": 201,
+    "message": "Account provisioned",
+    "code": "CREATED"
+  }
+}
+```
+
+Response with results:
+```
+{
+  "success": {
+    "status": 200,
+    "message": "Found 2 items",
+    "code": "OK"
+  },
+  "results": [
+    {
+      "account_id": 4,
+      "created_at": "2018-01-06T03:53:27.651Z",
+      "updated_at": "2018-01-06T03:53:27.651Z",
+      "account_number": null,
+      "sfdc_account_id": "3",
+      "id": 5
+    },
+    {
+      "account_id": 31,
+      "created_at": "2018-01-06T04:38:32.572Z",
+      "updated_at": "2018-01-06T04:38:32.572Z",
+      "account_number": null,
+      "sfdc_account_id": "1",
+      "id": 9
+    }
+  ]
+}
+```
+
+Response for get by id operation:
+```
+{
+  "success": {
+    "status": 200,
+    "message": "object found",
+    "code": "OK"
+  },
+  "results": {
+      "account_id": 4,
+      "created_at": "2018-05-06T03:53:27.651Z",
+      "updated_at": "2018-05-06T03:53:27.651Z",
+      "id": 5
+   }
+}
+```
+
+Response with results and service-defined results tag “rpz_hits”:
+```
+{
+  "success": {
+    "status": 200,
+    "message": "Read 360 items",
+    "code": "OK"
+  },
+  "rpz_hits": [
+    {
+      "pid": "default",
+      "rip": "10.35.205.4",
+      "policy_name": "Default",
+      "ttl": -1,
+      "qtype": 1,
+      "qip": "10.120.20.247",
+      "confidence": 3,
+      "network": "on-prem",
+      "event_time": "2017-12-13T07:07:50.000Z",
+      "feed_name": "rpz",
+      "dsource": 1,
+      "rcode": 3,
+      "timestamp": "11e7-dfd4-54e564f0-0000-0000287cd227",
+      "company": "302002|0",
+      "feed_type": "0",
+      "user": "unknown",
+      "device": "10.120.20.247",
+      "severity": 3,
+      "country": "unknown",
+      "policy_action": "Block",
+      "qname": "barfywyjgx.com",
+      "tproperty": "A",
+      "tclass": "UNKNOWN"
+    },
+    ...
+  ]
+}
+```
+
+### Errors
```

Comment:
Not sure how to deal with this as code related to errors is spread among two packages this one and `errors` package. Two readmes need to be linked somehow.

Reply:
Yeah, I agree. It's confusing.

Once we start building-out the package-level READMEs, it might make sense to rethink how I've categorized the packages.

```
- Request Handling Tools
- Server Utilities
- Database Utilities
- Testing
```

_Request Handling Tools_ is a bit vague. If we come up with more specific categories, it might clarify the distinction between the `rpc` and `errors` packages.